python/pyiceberg/catalog/rest.py

```diff
@@ -172,8 +175,6 @@ class OAuthErrorResponse(IcebergBaseModel):
 
 class RestCatalog(Catalog):
     uri: str
-    session: Session
```

Comment:
Based on my understanding, these should not be class-level attributes.  Changes to one set of properties or session should not affect other catalog instances.

Reply:
In Python it depends on the context:

By referencing it using `self`, it is bound to the object itself:

```python
class SomeClass:
    foo: str
    def __init__(self, foo: str):
        self.foo = foo

>>> class SomeClass:
...     foo: str
...     def __init__(self, foo: str):
...         self.foo = foo
... 
>>> a = SomeClass('a')
>>> a.foo
'a'
>>> b = SomeClass('b')
>>> b.foo
'b'
>>> a.foo
'a'  # still holds a
```

Class-level attributes are referenced by the class itself.
```python
>>> SomeClass.a = 'a'
>>> SomeClass.b = 'a'
>>> SomeClass.a
'a'
>>> SomeClass.b
'b'
>>> b.a
'a'
>>> SomeClass.c = 'c'
>>> a.c
'c'
```

I would suggest setting this to:

```python
    _session: Session
```

It is not strictly required, but it helps the developers to indicate which variables are available. The underscore `_` indicates that it is private to the class.

Reply:
I'll update since it appears we're using this elsewhere, but the behavior around this is very strange.

The class attribute needs to be initialized when declaring it as a class attribute. Without initialization, it doesn't exist:

```python
>>> class A:
...     attr: str
...     def print_attr(self):
...     print(self.attr)


>>> a = A()

>>> a.print_attr()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 a.print_attr()

Input In [14], in A.print_attr(self)
      3 def print_attr(self):
----> 4     print(self.attr)

AttributeError: 'A' object has no attribute 'attr'

>>> A.attr
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 A.attr

AttributeError: type object 'A' has no attribute 'attr'
```

If you initialize in the constructor, then it behaves the same as an instance attribute (not as a class attribute).  Alternatively, it you initialize at the class attribute level, then any assignment elsewhere will result in switching the class behavior to an instance behavior (which can produce some very strange behavior).

Reply:
Yes the `attr: str` in:

```python
class A:
    attr: str
```

is just syntactical sugar, and will not be there until it is assigned. If you assign a value to it directly, it won't throw an `AttributeError`.

```
➜  trino git:(master) ✗ python3
Python 3.11.3 (main, Apr  7 2023, 20:13:31) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class A:
...     attr: str = ""
... 
>>> A().attr
''
```